gems/sorbet-runtime/test/types/final_method.rb

```diff
@@ -0,0 +1,272 @@
+  end
+
+  it "forbids redefining a final class method with no sig" do
+    err = assert_raises(RuntimeError) do
+      Class.new do
+        extend T::Sig
+        sig(:final) {void}
+        def self.foo; end
+        def self.foo; end
+      end
+    end
+    assert_includes(err.message, "was declared as final and cannot be redefined")
+  end
+
+  it "allows redefining a regular instance method to be final" do
+    Class.new do
+      extend T::Sig
+      def foo; end
+      sig(:final) {void}
+      def foo; end
+    end
+  end
+
+  it "allows redefining a regular class method to be final" do
+    Class.new do
+      extend T::Sig
+      def self.foo; end
+      sig(:final) {void}
+      def self.foo; end
+    end
+  end
+
+  it "forbids overriding a final instance method" do
+    c = Class.new do
+      extend T::Sig
+      sig(:final) {void}
+      def foo; end
+    end
+    err = assert_raises(RuntimeError) do
+      Class.new(c) do
+        def foo; end
+      end
+    end
+    assert_includes(err.message, "was declared as final and cannot be overridden")
+  end
+
+  it "forbids overriding a final class method" do
+    c = Class.new do
+      extend T::Sig
+      sig(:final) {void}
+      def self.foo; end
+    end
+    err = assert_raises(RuntimeError) do
+      Class.new(c) do
+        def self.foo; end
+      end
+    end
+    assert_includes(err.message, "was declared as final and cannot be overridden")
+  end
+
+  it "forbids overriding a final method from an included module" do
+    m = Module.new do
+      extend T::Sig
+      sig(:final) {void}
+      def foo; end
+    end
+    err = assert_raises(RuntimeError) do
+      Class.new do
+        include m
+        def foo; end
+      end
+    end
+    assert_includes(err.message, "was declared as final and cannot be overridden")
+  end
+
+  it "forbids overriding a final method from an extended module" do
+    m = Module.new do
+      extend T::Sig
+      sig(:final) {void}
+      def foo; end
+    end
+    err = assert_raises(RuntimeError) do
+      Class.new do
+        extend m
+        def self.foo; end
+      end
+    end
+    assert_includes(err.message, "was declared as final and cannot be overridden")
+  end
+
+  it "forbids overriding a final method by including two modules" do
+    m1 = Module.new do
+      extend T::Sig
+      sig(:final) {void}
+      def foo; end
+    end
+    m2 = Module.new do
+      def foo; end
+    end
+    err = assert_raises(RuntimeError) do
+      Class.new do
+        include m1, m2
+      end
+    end
+    assert_includes(err.message, "was declared as final and cannot be overridden")
+  end
+
+  it "forbids overriding a final method by extending two modules" do
+    m1 = Module.new do
+      extend T::Sig
+      sig(:final) {void}
+      def foo; end
+    end
+    m2 = Module.new do
+      def foo; end
+    end
+    err = assert_raises(RuntimeError) do
+      Class.new do
+        extend m1, m2
```

Comment:
This diff makes your test fail:

```diff
❯ gd
diff --git a/gems/sorbet-runtime/test/types/final_method.rb b/gems/sorbet-runtime/test/types/final_method.rb
index eace3a694..5bd1214a8 100644
--- a/gems/sorbet-runtime/test/types/final_method.rb
+++ b/gems/sorbet-runtime/test/types/final_method.rb
@@ -202,6 +202,7 @@ class Opus::Types::Test::FinalMethodTest < Critic::Unit::UnitTest
     end
     err = assert_raises(RuntimeError) do
       Class.new do
+        def self.extend; end
         extend m1, m2
       end
     end
```

Which means that all I have to do to disable final checks is to add this line to my class. Can we avoid this?

Reply:
Maybe we could make extend itself a final method? But why would someone redefine extend? This prevents you from being able to extend things. Like, the error here happens without any of the sorbet-runtime machinery:

```rb
Class.new do
  def self.extend; end
  extend Module.new
end
```

Similar error happen when you try to extend a module that has overridden `extends` to, for example, take no arguments.

Because ruby essentially lets you "do anything" at runtime (redefine methods, step around method privacy, etc), I feel like no matter how hard we try to prevent people from doing "bad" things, there will be a case where a sufficiently determined user could foil sorbet-runtime. The question is, where do we draw the line for, "although we can't prevent the user from stepping around us in this certain way, it's an uncommon enough pattern that we don't want to support it"?

The things that I can think of that, for me, both break the final checks but are also uncommon enough that I'm not sure if we should care are: redefining `include`, `extend`, `included`, or `extended` in a "strange way" (takes no arguments, doesn't super, etc). This includes your example.

Reply:
Resolved offline: We're not going to put extra effort into protecting against this particular footgun. We're going to depend on users, if they are to define their own `included`, etc, to be good citizens by accepting the right number of arguments and calling `super` as appropriate.